### DISCUSSION TITLE
Support for editable tables

### DISCUSSION THREAD

**Discussion by xiaowangsmile** [fabricjs/fabric.js#8310] [Ideas]
Hello, everyone
Fabric. Js is a great project, and my current project is based on it,Tables are a very important module for me；
I'm going to draw on the DOM first,Then get the DOM raw data to draw and group on the canvas，When changes  again （eg.Merge cells     add or delete rows）, I need to continue feeding the canvas data back into the DOM for editing;Experience is not good，Is there a plan to support tables module  in a future release？ thank you

**Comment by ShaMan123** [comment]
I would like to suggest a few things:
1. Make sure you work with the new group #7670 , it is built for such customizations
2. Use the controls API to do all UX
3. I wouldn't use a html element, but that's me

It is not on our plan AFAIK. 
v6 in designed for easy subclassing so the community can build such an object (you perhaps?) and then use it by a simple import as any library handles extending.

**Reply by shivasubramaniyam** [reply]
Is there any ongoing or planned development for native table support in Fabric.js (e.g., editable tables, rows/columns, resizing)?

**Comment by mauricekindermann** [comment]
I'd love a simple FigJam/Canva style table editor too. Last month I made a quick related experiment: https://github.com/fabricjs/fabric.js/discussions/10819

Finished it a few weeks later and called it Layout:

<img width="818" height="491" alt="image" src="https://github.com/user-attachments/assets/db41e65b-a5aa-4724-b261-b15a5d0681a8" />

It uses a LayoutStrategy with a few custom properties, but it's pretty much just a fancy edtibale group, the v6 group rewrite made it possible. I disabled Layout nesting - the edge cases were too complex for my needs, but I can imagine how a similar approach would work for Tables.

If I get to it I'll PR my Layout as a FabricExtension, just gotta make it portable and more generic for most peopels use cases. If that goes well I can see tables on the cards.

For future reference....

This discussion had a few insights: https://github.com/fabricjs/fabric.js/issues/4434

Including this post: 
https://shorterloop.com/the-product-mindset/posts/create-dynamic-sections-with-drag-and-drop-functionality-and-customizable-textboxes/

I also did some research a few weeks ago and found some related useful discussions on tldraw and excalidraw
https://github.com/tldraw/tldraw/issues/1780
https://github.com/excalidraw/excalidraw/issues/4847
https://cliftonc.nl/blog/adding-tables-to-tldraw-with-tiptap/

DrawIO has a Graph.js class with some homework to copy `Graph.prototype.createTable(rowCount, colCount, w, h, title, ...)`
https://github.com/jgraph/drawio/blob/dev/src/main/webapp/js/grapheditor/Graph.js

PenPot also has some related code
https://github.com/penpot/penpot/tree/develop/common/src/app/common/geom/shapes/grid_layout

And some raw canvas examples
https://www.npmjs.com/package/@borderliner/canvas-table
https://github.com/el/canvas-table

I like to see how everyone else does it first before I take a stab at things to avoid reinventing the wheel.

**Comment by mauricekindermann** [comment]
I couldn't help myself and made a proof of concept:
https://codepen.io/goldencarers/pen/MYepgNw

- ClickDrag to move
- ClickCell to change color
- DoubleClick to edit text
- Text changes reflow Table
- Cell width/height/padding/spacing
- Table/Stroke/Cell fill
- Shift MultiCell selection (eg A1 - C3, C1 - C5)
- ESC to backout of Text > Cell > Table
- ActiveCell ctx border (prevent transient pollution)
- Copy/Paste between cells
- Keyboard cell navigation

Todo / Untested
- Disable Flip (too troublesome I fear)
- Merge cells (yikes)
- Drag borders to resize columns/rows (with min width/height?)
- Custom corner controls for add/remove (+inherit adjacent styles?)
- Support scaled table font size
- MultiCellSelect edits
- Fix broken Table size changes during text editing

All the fabric building blocks are there, its just orchestration + custom behaviour. I'm not sure what a Table class would even look like in fabric, so for now I think this makes sense as a future extension.

Pretty neat!

**Reply by pibolito** [reply]
Hello Mauriece! First let me congratulate you for the fabrice.js table project! It seems really awesome! I was looking for something similar but with a line above the table for the title! It has been giving me lots of headaches and sleeping hours lost just to accomplish that :\
I was wondering if you could consider making a version with that funcionality?
Thank you very much in advance!
pib

**Comment by shivasubramaniyam** [comment]
Hey @mauricekindermann👋,
Thank you for your support and I really appreciate it.

I have gone through your POC, looks amazing to me.

You have the table implementation with more customization features.

While I was waiting for your reply, I have tried a few things in table implementation.

[FabricTable.js](https://github.com/user-attachments/files/24676306/FabricTable.js)
[tables-utils.js](https://github.com/user-attachments/files/24676309/tables-utils.js)

this would help you to add the resizing the table to your POC

thank you,
happy coding..

**Reply by mauricekindermann** [reply]
Nice, I think I've got the rest all done tonight. Sharing shortly.

**Comment by mauricekindermann** [comment]
I think i'm done:
https://codepen.io/goldencarers/pen/MYepgNw

New features: 
- resize cols/rows
- multi cell edits
- merge cells
- calc text height
- merge text cells
- handle scaled tables
- custom edge handles
- utilize more Fabric APIs
- smart ctx escapes
- cascading table>cell fill/border

At this point I'm not sure what else I would add that is not domain-specific. Not bad at all, fabric rocks. 

I just need to clean up the code, organize it properly, add a bunch of tests, then PR it into the official extensions.

**Reply by xiaowangsmile** [reply]
This is really amazing. I never expected that a requirement I raised back in 2022 would receive a response now—and be implemented so well. I’ve tested your demo, and it works wonderfully.

I’d like to share a couple of suggestions:
First, when merging cells, would it be possible to select multiple cells directly with the mouse, without needing to use the keyboard?
Second, I noticed that when I enter text into a merged cell, its height changes. It would be great if the height could remain unchanged as long as the text does not wrap to a new line.

You may refer to Canva (https://www.canva.com/
) for similar behavior and implementation ideas.

Thank you very much!

**Reply by mauricekindermann** [reply]
No worries. Interestingly, those 2 particular issues should already be working:

1. Click + Shift + Click to select then merge  (though I didn't do CMD click)
2. Text should reflow (copy paste, change font size, scale, cell padding, col/row resize, etc) but I might not have caught that one as merged cells was my last feature 

I also fixed serialization for independent row/col width.

Look I have no doubt there are many unhandled edge cases as I haven't thoroughly tested yet, I just wanted to see if it was possible. 

I'm happy to fix any issues I can easily replicate, just help me reproduce them step by step. In fact I'd love as much testing and problems found as possible because I'll be putting this in a production app eventually. 

My plan is to PR this as a proper nicely organized extension, then I'd like to get it working well with AlignGuidelines (and improve that extension as well to add edge scale guidelines).

**Reply by shivasubramaniyam** [reply]
Thanks for the great support.
These implementations have made it much easier to add features to the canvas.

I was wondering if there are any other features you’ve already implemented that would be useful for me to look into.

:)

**Reply by mauricekindermann** [reply]
I have a tone of stuff but none that is shareable in any useful way. Hopefully I can contribute much of it.

**Reply by mauricekindermann** [reply]
What are you most interested in? I was planning PenTool, PathEditing, or maybe Layout next. 

In terms of architecture - TLDraw have a lot of goodies you can source dive, but you have to adapt it to Fabric and your stack. 

My main managers: 

- AlignmentManager
- AnimationManager
- ArrowManager
- CanvasManager
- ClipboardManager
- ColorStyleManager
- CropManager
- DistributeManager
- DraftManager
- DrawingManager
- DropManager
- ExportManager
- EyeDropperManager
- FillManager
- FontManager
- GradientManager
- GroupManager
- HistoryManager
- ImageFiltersManager
- ImageManager
- InteractionManager
- KeyboardManager
- LayoutManager
- LockedObjectManager
- MaskManager
- ObjectControlsManager
- ObjectInteractionManager
- ObjectPropertiesManager
- ObjectStyleManager
- PageManager
- PanningManager
- PaperManager
- PathManager
- PointEditingManager
- SelectionManager
- ShadowManager
- SnapManager
- TextStyleManager
- ToolbarCapabilityManager
- ViewportManager
- ZoomManager
- TextOnPathManager

**Reply by asturur** [reply]
pageObject / ruler / gridsnap is something i was thinking to add on my own.

**Reply by mauricekindermann** [reply]
Oh nice, page object I'd love to see how you implement that. Is your page object similar to a canva page background, or a figma frame, or something else entirely?

**Reply by asturur** [reply]
is something i have done for older projects, is just a class that you can use as a reference to understand how large your document is, and that provides reference for creating a sort of dark layer outside the page, so that you have the impression the canvas finishes but the controls are visible outside of it.
Is also useful to bind the viewport manipulation and find the correct zoom level, also supports checkerboard to represent opacity, don't let the pan escape page out of view...
Is really just a fancy rect and some utilities for viewport.
As usual a guideline on how you can build thing.

**Reply by mauricekindermann** [reply]
Ha! Sounds very simliar to what I built but with cooler names. 

I call mine the backgroundRect, then I use a ctx pageOverlay to do the same control passthrough outside boundaries you described. The bgRect size is stored next to seralized content (and generated when you load rather than part of the objects). But I ran into lots of situations where it would get manipulated so I ended up creating a lot of guards, like `ensureBackgroundLayering`, `setupBackgroundProtection`, and special seralization handling. 

I'd love to see what you built and how you handled it.

**Reply by asturur** [reply]
What i built is buried in a dead app and was tied to the app, i just want to make it a bit modular and packaged, so it can be used or at least learned.

**Reply by mauricekindermann** [reply]
Nice I'm very keen to see. Rulers is one I almost did myself, but its not super important for what i'm building, and I figured there are a lot more papercuts than I imagine there are (there always is with transietn states, seralization exclusion, zoom/pan positioning, locking, creation UI, disposa UXl, etc etc).

**Comment by mauricekindermann** [comment]
Submitted a PR to include this as an official extension: https://github.com/fabricjs/fabric.js/pull/10852

Fixed lots of bugs, added lots of bells and whistles. 

Clone down the PR and run `npm start vanilla` to play with it.

**Comment by shivasubramaniyam** [comment]
I’ve implemented drawing tools including a pen, brush, highlighter, and eraser. However, the eraser only works correctly on a white background. When the background color is changed, the erased areas appear white instead of transparent.

is there any other approach to it ??

**Reply by mauricekindermann** [reply]
Start a discussion, provide code samples, screenshot, and ideally an MVP working sample. Can't debug otherwise

**Reply by shivasubramaniyam** [reply]
In this image attached,
I have drawn my name and the erased few area
i tried changing the bg colour of canvas to red 
so the erased area is white now.
ideally erased area should match the bg..


<img width="733" height="524" alt="image" src="https://github.com/user-attachments/assets/b281c18f-b260-4031-93c5-21cbc4fc79dd" />

[cust.txt](https://github.com/user-attachments/files/25044137/cust.txt)

**Reply by mauricekindermann** [reply]
Could you post a new discussion? Then I can answer there, this is not relevant to tables anymore.
